### DISCUSSION TITLE
Unit testing? - CLOSED

### DISCUSSION THREAD

**Discussion by eharrow** [astlouys/Pico-Green-Clock#13] [Ideas]
Unfortunately I'm not a C developer other than a few weeks at university.  I do develop in  java, javascript, some python and a smattering of rust and would like to do try my hand with this project a bit more.

Do you have any thoughts to how to unit test the code i.e. without it hooked up to the clock.  Perhaps how to debug with a pico outside the clock mocking the display, buttons, rtc and buzzer would be interesting.  No idea how possible this is with embedded boards like this but in the enterprise world I develop for we do make use of mocking frameworks with test harnesses.

**Comment by astlouys** [comment]
Hi Ewan,

Obviously, having other users making tests is definitely a good thing !  The more problems that can be found, the better becomes the application. Moreover, there has been so many features / options added to the clock in this project, I must confess that sometimes, I was quick on the tests as I had so many other functions to develop (don't follow my example !).

However, in the case of the Green Clock, I believe there are too many close links between the hardware and the software to be able to make really valuable unit tests. Problems may be related to the display and / or the scrolling (that you won't see if you don't have the clock), there is also the real-time clock IC keeping the time, the flash keeping the configuration, plus timings that may depend on some hardware behaviour etc... etc...

If you want to make a first step in the C language and using the "environment" of the Green Clock, I think the farther you could go would be to take come specific functions in the source code and replicate them in another program that could be used in "stand-alone" mode on another Pico. For example, the Daylight Saving Time / Summer Time is something that has been discussed in the recent weeks. It could be something that you extract from the Green Clock code and test on another Pico to see if all the logic is OK.

There are probably other parts of the code that can be extracted like this, but I would stay away of things related to the Green Clock hardware... or you buy another Green Clock that you reserve for testing purposes ;-)

Hope this help...
